### ISSUE TITLE
client, consent: rethink pagination

### ISSUE THREAD

**Issue by lopezator** [ory/hydra#1434]
**Is your feature request related to a problem? Please describe.**

The pagination functionality inside client and consent packages is being made using a "count(*)", this could lead to performance problems on big sets of data.

see:

https://github.com/ory/hydra/blob/f8f23630d80f8980786ef85a49597f9b6b3eec7b/client/manager_sql.go#L352-L360

and:

https://github.com/ory/hydra/blob/f8f23630d80f8980786ef85a49597f9b6b3eec7b/consent/manager_sql.go#L474-L490

**Describe the solution you'd like**

"Guys, we’re doing pagination wrong…":

https://hackernoon.com/guys-were-doing-pagination-wrong-f6c18a91b232

"Faster Pagination in Mysql – Why Order By With Limit and Offset is Slow?":

https://www.eversql.com/faster-pagination-in-mysql-why-order-by-with-limit-and-offset-is-slow/

"Web API Pagination with the 'Timestamp_Offset_Checksum' Continuation Token":

https://phauer.com/2017/web-api-pagination-continuation-token/

**Describe alternatives you've considered**

In summary, a token-based pagination solution.

**Comment by aeneasr** [comment]
This was a conscious trade-off. Implementing token-based ordering is (to my knowledge) heavily dependent on the underlying db technology. Since hydra is (essentially) db agnostic it would be some work to get that done. For the consent case specifically, it is very unlikely to have more than a couple of consent requests that match that query. Keep in mind that filtering comes first before actually counting.

For OAuth2 Clients - unless you're operating at Google scale, it's also unlikely to have that many clients that this query could become slow.

I don't think this will be fixed in the near future unless some prod system actually encounters a problem here.

**Comment by lopezator** [comment]
I disagree on the db-dependent argument. We are using internally on our API a token-based db-independent solution.

I take the point for the number of rows, I missed the filter-before-count logic.

I close this for now then, we can always reopen it if this turns to be a problem to someone in the future.

Thank you very much for your response!
